addons/account/models/account_account.py

```diff
@@ -685,18 +790,48 @@ def _compute_display_name(self):
             account.display_name = f"{account.code} {account.name}"
 
     def copy_data(self, default=None):
-        default = dict(default or {})
         vals_list = super().copy_data(default)
-        cache_map = defaultdict(set)
+        default = default or {}
+
+        # This makes the behaviour of env['account.account'].copy() dependent on the active company:
+        # When copying an account, the new account is created in the current active company, unless a different
+        # company is specified in the defaults. This allows users to copy accounts that belong to multiple
+        # companies despite the limitation that a new account can only be created belonging to a single company.
+        match default.get('company_ids'):
+            case None:
+                company = self.env.company
+            case [(Command.LINK, company_id, 0)] | [(Command.SET, 0, [company_id])] | [company_id] if isinstance(company_id, int):
+                company = self.env['res.company'].browse(company_id)
+            case _:
+                raise ValueError(_("You may only give an account a single company at creation."))
+
+        # We must restrict check_company fields to values available to the company of the new account.
+        fields_to_filter_by_company = {
+            field
+            for field in self.env['ir.model.fields'].search([
+                ('model', '=', 'account.account'),
+                ('ttype', 'in', ('many2many', 'one2many')),
+                ('store', '=', True),
+            ])
+            if self.env[field.model]._fields[field.name].check_company
+        }
+        cache = set()
         for account, vals in zip(self, vals_list):
             if 'code' not in default:
-                company = default.get('company_id', account.company_id)
-                company = company if isinstance(company, models.BaseModel) else account.env['res.company'].browse(company)
-                cache = cache_map[company.id]
-                vals['code'] = account._search_new_account_code(account.code, company, cache)
+                start_code = account.code or account.with_company(account.company_ids[0]).code
```

Comment:
I agree about using `with_company(company)`, but I think we still need to fallback on `account.with_company(account.company_ids[0]).code` in case the code isn't set for `company`...